modules/ppcp-button/resources/js/modules/Renderer/PreviewButton.js

```diff
@@ -0,0 +1,135 @@
+import merge from "deepmerge";
+
+/**
+ * Base class for APM button previews, used on the plugin's settings page.
+ */
+class PreviewButton {
+    /**
+     * @param {string} selector - CSS ID of the wrapper, including the `#`
+     * @param {object} apiConfig - PayPal configuration object; retrieved via a
+     * widgetBuilder API method
+     */
+    constructor({selector, apiConfig}) {
+        this.apiConfig = apiConfig;
+        this.defaultAttributes = {};
+        this.buttonConfig = {};
+        this.ppcpConfig = {};
+        this.isDynamic = true;
+
+        // The selector is usually overwritten in constructor of derived class.
+        this.selector = selector;
+        this.wrapper = selector;
+
+        this.domWrapper = null;
+    }
+
+    /**
+     * Creates a new DOM node to contain the preview button.
+     *
+     * @return {jQuery} Always a single jQuery element with the new DOM node.
+     */
+    createNewWrapper() {
+        const previewId = this.selector.replace('#', '')
+        const previewClass = 'ppcp-button-apm';
+
+        return jQuery(`<div id="${previewId}" class="${previewClass}">`)
+    }
+
+    /**
+     * Toggle the "dynamic" nature of the preview.
+     * When the button is dynamic, it will reflect current form values. A static button always
+     * uses the settings that were provided via PHP.
+     *
+     * @return {this} Reference to self, for chaining.
+     */
+    setDynamic(state) {
+        this.isDynamic = state;
+        return this;
+    }
+
+    /**
+     * Sets server-side configuration for the button.
+     *
+     * @return {this} Reference to self, for chaining.
+     */
+    setButtonConfig(config) {
+        this.buttonConfig = merge(this.defaultAttributes, config)
+        this.buttonConfig.button.wrapper = this.selector
+
+        return this;
+    }
+
+    /**
+     * Updates the button configuration with current details from the form.
+     *
+     * @return {this} Reference to self, for chaining.
+     */
+    setPpcpConfig(config) {
+        this.ppcpConfig = merge({}, config);
```

Comment:
Good catch 👍 The idea of this part is, to create a local copy of the copy.

Since the config object can be modified outside the class, we want to keep a static version of the provided configuration.

```js
this.setPpcpConfig(config)
...
config.color = 'black'
```

Without the "merge", this would cause some unexpected behavior; with "merge" changing the config object does not impact the preview, until explicitly calling `setPpcpConfig()` again.